x/stakeibc/keeper/icacallbacks.go

```diff
@@ -116,3 +118,43 @@ func DelegateCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack
 	k.RecordsKeeper.RemoveDepositRecord(ctx, cast.ToUint64(recordId))
 	return nil
 }
+
+// ----------------------------------- redemption callback ----------------------------------- //
+func (k Keeper) MarshalRedemptionCallbackArgs(ctx sdk.Context, redemptionCallback types.RedemptionCallback) []byte {
+	out, err := proto.Marshal(&redemptionCallback)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("MarshalRedemptionCallbackArgs %v", err.Error()))
+	}
+	return out
+}
+
+func (k Keeper) UnmarshalRedemptionCallbackArgs(ctx sdk.Context, redemptionCallback []byte) types.RedemptionCallback {
+	unmarshalledDelegateCallback := types.RedemptionCallback{}
+	if err := proto.Unmarshal(redemptionCallback, &unmarshalledDelegateCallback); err != nil {
+        k.Logger(ctx).Error(fmt.Sprintf("UnmarshalRedemptionCallbackArgs %v", err.Error()))
+	}
+	return unmarshalledDelegateCallback
+}
+
+func RedemptionCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack *channeltypes.Acknowledgement_Result, args []byte) error {
+	// QUESTION: should we check invariants here? e.g. sendMsg.FromAddress == redemptionAddress, msg type == MsgSend, etc.
+	k.Logger(ctx).Info("RedemptionCallback executing", "packet", packet, "ack", ack, "args", args)
+	// deserialize the args
+	redemptionCallback := k.UnmarshalRedemptionCallbackArgs(ctx, args)
+	k.Logger(ctx).Info(fmt.Sprintf("RedemptionCallback %v", redemptionCallback))
+	userRedemptionRecord, found := k.RecordsKeeper.GetUserRedemptionRecord(ctx, redemptionCallback.GetUserRedemptionRecordId())
+	if !found {
+		return sdkerrors.Wrap(types.ErrRecordNotFound, "user redemption record not found")
+	}
+
+	if ack == nil {
+		// transaction on the host chain failed
+		// set UserRedemptionRecord as claimable
+		userRedemptionRecord.IsClaimable = true
+		k.RecordsKeeper.SetUserRedemptionRecord(ctx, userRedemptionRecord)
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "ack is nil")
```

Comment:
do we get any useful info back on failure?

Reply:
not really... although I'll log the packet